deploy/unifi.sh

```diff
@@ -42,102 +55,123 @@ unifi_deploy() {
+    # shellcheck disable=SC2181
+    if [ "$?" != "0" ]; then
+      _err "Oops, error creating import pkcs12, please report bug to us."
       return 1
+    fi
+
+    _debug "Import into keystore: $_unifi_keystore"
+    if keytool -importkeystore \
+      -deststorepass "$_unifi_keypass" -destkeypass "$_unifi_keypass" -destkeystore "$_unifi_keystore" \
+      -srckeystore "$_import_pkcs12" -srcstoretype PKCS12 -srcstorepass "$_unifi_keypass" \
+      -alias unifi -noprompt; then
+      _debug "Import keystore success!"
+      rm "$_import_pkcs12"
     else
-      _err "It seems that the specified unifi keystore is not valid, please check."
+      _err "Error importing into Unifi Java keystore."
+      _err "Please re-run with --debug and report a bug."
+      rm "$_import_pkcs12"
       return 1
     fi
-  fi
-  if [ ! -w "$_unifi_keystore" ]; then
-    _err "The file $_unifi_keystore is not writable, please change the permission."
+
+    DEFAULT_UNIFI_RELOAD="${DEFAULT_UNIFI_RELOAD} ${DEFAULT_UNIFI_RELOAD:+&&} service unifi restart"
+    _info "Install Unifi Controller certificate success!"
+  elif [ "$DEPLOY_UNIFI_KEYSTORE" ]; then
+    _err "The specified DEPLOY_UNIFI_KEYSTORE='$DEPLOY_UNIFI_KEYSTORE' is not valid, please check."
     return 1
   fi
 
-  _debug _cloudkey_deploy "$_cloudkey_deploy"
-  _debug _cloudkey_certdir "$_cloudkey_certdir"
-  if [ "$_cloudkey_deploy" = "yes" ]; then
-    if [ ! -d "$_cloudkey_certdir" ]; then
-      if [ -z "$DEPLOY_UNIFI_CLOUDKEY_CERTDIR" ]; then
-        _err "Attempting to deploy on a Cloud Key, but cert dir not found in usual location."
-        _err "Either set DEPLOY_UNIFI_CLOUDKEY=no or define DEPLOY_UNIFI_CLOUDKEY_CERTDIR."
-        return 1
-      else
-        _err "The Cloud Key cert dir $_cloudkey_certdir is not valid, please check."
-        return 1
-      fi
-    fi
+  # Cloud Key Gen1 environment (nginx admin pages) --
+  # auto-detect by file /etc/ssl/private/cloudkey.key:
+  DEFAULT_DEPLOY_UNIFI_CLOUDKEY_CERTDIR="/etc/ssl/private"
+  _cloudkey_certdir="${DEPLOY_UNIFI_CLOUDKEY_CERTDIR:-$DEFAULT_DEPLOY_UNIFI_CLOUDKEY_CERTDIR}"
+  if [ -f "${_cloudkey_certdir}/cloudkey.key" ]; then
+    _info "Installing certificate for Cloud Key Gen1 (nginx admin pages)"
+    _debug _cloudkey_certdir "$_cloudkey_certdir"
     if [ ! -w "$_cloudkey_certdir" ]; then
       _err "The directory $_cloudkey_certdir is not writable; please check permissions."
       return 1
     fi
-  fi
 
-  _info "Generate import pkcs12"
-  _import_pkcs12="$(_mktemp)"
-  _toPkcs "$_import_pkcs12" "$_ckey" "$_ccert" "$_cca" "$_unifi_keypass" unifi root
-  if [ "$?" != "0" ]; then
-    _err "Oops, error creating import pkcs12, please report bug to us."
+    cp "$_cfullchain" "${_cloudkey_certdir}/cloudkey.crt"
+    cp "$_ckey" "${_cloudkey_certdir}/cloudkey.key"
+    (cd "$_cloudkey_certdir" && tar -cf cert.tar cloudkey.crt cloudkey.key unifi.keystore.jks)
```

Comment:
This `tar` command is adding the keystore file - but it's not copied or generated in place like it is above for a self-hosted controller. I wonder if the java keystore should be generated regardless in the `$domain` dir, and just copied into place when required? (either into `$DEFAULT_UNIFI_KEYSTORE`, or `$DEFAULT_DEPLOY_UNIFI_CLOUDKEY_CERTDIR/unifi.keystore.jks` ?

Reply:
(this location on the CK is where the SDN controller loads its cert from)

Reply:
On my CK Gen1, /usr/lib/unifi/data/keystore is a symlink to /etc/ssl/private/unifi.keystore.jks. I'm assuming that's the standard setup? (And that this is how the current released version of the unifi deploy-hook is able to partially work on a Cloud Key.)

Is there a valid CK config where /usr/lib/unifi/data/keystore is *different* from /etc/ssl/private/unifi.keystore.jks, that this deploy hook should support? If not, it might be easiest to just check before generating the tar that those files have the same contents, and error with "unsupported CloudKey config" if they don't.

(My concern with generating the java keystore in $domain and copying it into place is that the released unifi hook *imports* into the existing keystore. Copying over it would blow away any other certs that might be in there, right? Not a concern on a CK, but I don't know what people might be doing with self-hosted Controller configs.)

Reply:
Added this before the tar, see if you think that's sufficient:

```sh
    # Cloud Key expects to load the keystore from /etc/ssl/private/unifi.keystore.jks.
    # Normally /usr/lib/unifi/data/keystore is a symlink there (so the keystore was
    # updated above), but if not, we don't know how to handle this installation:
    if ! cmp -s "$_unifi_keystore" "${_cloudkey_certdir}/unifi.keystore.jks"; then
      _err "Unsupported Cloud Key configuration: keystore not found at '${_cloudkey_certdir}/unifi.keystore.jks'"
      return 1
    fi
```

Reply:
OK, a bit to disect here.

> On my CK Gen1, /usr/lib/unifi/data/keystore is a symlink to /etc/ssl/private/unifi.keystore.jks. I'm assuming that's the standard setup? (And that this is how the current released version of the unifi deploy-hook is able to partially work on a Cloud Key.)

Well, look at that - it is too. I was aware of `/srv/unifi/data` on the CK, but did not know that's where `/usr/lib/unifi/data` is linked to:

```
root@CK:~# ls -l /usr/lib/unifi/data
lrwxrwxrwx 1 root root 15 Jan  4 20:19 /usr/lib/unifi/data -> /srv/unifi/data
root@CK:~# ls -l /srv/unifi/data/keystore
lrwxrwxrwx 1 unifi unifi 35 Apr 16  2020 /srv/unifi/data/keystore -> /etc/ssl/private/unifi.keystore.jks
```

> Is there a valid CK config where /usr/lib/unifi/data/keystore is _different_ from /etc/ssl/private/unifi.keystore.jks, that this deploy hook should support? If not, it might be easiest to just check before generating the tar that those files have the same contents, and error with "unsupported CloudKey config" if they don't.

I am not familiar enough with the internals of the SDN controller to say for sure, but you are probably more than safe doing that.

> (My concern with generating the java keystore in $domain and copying it into place is that the released unifi hook _imports_ into the existing keystore. Copying over it would blow away any other certs that might be in there, right? Not a concern on a CK, but I don't know what people might be doing with self-hosted Controller configs.)

Fair point. I'd actually completely missed that it was importing - with my previous certificate management (before upgrading to UnifiOS), I was actually generating a new keystore on a remote host, and scp'ing it into place. What you're doing seems sane.

> Added this before the tar, see if you think that's sufficient:

Yes, more than sufficient. Because you're going to be running on a debian system, or a CK, you could also assume the existence of `realpath`, so you could check its output instead (or as well as):

```
root@CK:~# realpath /usr/lib/unifi/data/keystore
/etc/ssl/private/unifi.keystore.jks
root@CK:~# [ "$(realpath "$_unifi_keystore")" = "${_cloudkey_certdir}/unifi.keystore.jks" ] && echo true
true
root@CK:~#
```

But, at a guess, I'd say that the controller only uses `/usr/lib/unifi/data/keystore` at startup, so what's in `/etc/ssl/private` is probably irrelevant if it's not the ultimate location of the symlink.

In summary, what you have is great. good stuff.

Reply:
OK. FWIW, there's a [forum comment from UI-Team](https://community.ui.com/questions/Unifi-Cloud-key-certificate-installation/d3483712-e44c-4d94-a0e0-c50093b5bd09#answer/03a79172-0b78-4539-be29-7ff1a39e883d) about the files in /etc/ssl/private, that the CK "boot up process will check if they are out of sync" with the cert.tar contents. (But it doesn't say what happens if they are.)